include/dgl/sampler.h

```diff
@@ -61,6 +90,54 @@ class SamplerOp {
                             IdArray seeds,
                             int num_traces,
                             int num_hops);
+
+  /*!
+   * \brief Batch-generate random walk traces with restart
+   *
+   * Stop generating traces if max_frequrent_visited_nodes nodes are visited more than
+   * max_visit_counts times.
+   *
+   * \param seeds The array of starting vertex IDs
+   * \param restart_prob The restart probability
+   * \param max_nodes_per_seed Stop generating traces if this many nodes are visited for each seed
+   * \param num_hops The number of hops for each trace
+   * \return a flat ID array with shape (num_seeds, num_traces, num_hops + 1)
+   */
+  static RandomWalkTraces RandomWalkWithRestart(
+      const GraphInterface *gptr,
+      IdArray seeds,
+      double restart_prob,
+      uint64_t max_nodes_per_seed,
+      uint64_t max_visit_counts,
+      uint64_t max_frequent_visited_nodes);
+
+  /*
+   * \note Doesn't verify whether the graph is indeed a bipartite graph
+   */
+  static RandomWalkTraces BipartiteSingleSidedRandomWalkWithRestart(
+      const GraphInterface *gptr,
+      IdArray seeds,
+      double restart_prob,
+      uint64_t max_nodes_per_seed,
+      uint64_t max_visit_counts,
+      uint64_t max_frequent_visited_nodes);
+
+ private:
+  static IdArray GenericRandomWalk(
+      const GraphInterface *gptr,
+      IdArray seeds,
+      int num_traces,
+      int num_hops,
+      bool (*walker)(const GraphInterface *, unsigned int *, dgl_id_t, dgl_id_t *));
+
+  static RandomWalkTraces GenericRandomWalkWithRestart(
+      const GraphInterface *gptr,
+      IdArray seeds,
+      double restart_prob,
+      uint64_t max_nodes_per_seed,
+      uint64_t max_visit_counts,
+      uint64_t max_frequent_visited_nodes,
+      bool (*walker)(const GraphInterface *, unsigned int *, dgl_id_t, dgl_id_t *));
```

Comment:
I changed it to `using Walker = std::function<...>` (`randomwalk.cc:25`).

Reply:
This will not support structure functor:
```c++
struct Functor {
  bool operator() (...) { ... }
};
```

Reply:
Works for me:
```c++
#include <functional>
#include <iostream>
struct A {
  int c;
  int operator () (int a, int b) { return a + b + c; };
};

int main()
{
  A a;
  a.c = 3;
  std::function<int(int, int)> f = a;
  std::cout << f(2, 4) << std::endl;
  return 0;
}
```

Reply:
It works on my side too. However, I have a pretty new GCC (8.2.1) so I don't know whether this works for older version and whether this is a new feature included in C++17 and later. Is this in standard? I searched a little bit but cannot find an answer, and there are many posts about C++ failed to convert even lambda to function. I would say template is still a safer approach.

Reply:
Another (small) benefit of using template is no explicit signature is required. For example, we could free to pass function with `const` or `const &` type argument types.

Reply:
Re GCC: Passed on GCC 4.8.2 + CentOS 6 (which is pretty old).  Also it is said that `std::function` can represent any callable.
Re Signature: I would prefer explicit signature though, especially since the code calling `walker` *explicitly* specifies the arguments anyway.

Reply:
Ok. Just make sure it works for MSVC and Clang as well.